### PR TITLE
check rw_counter is unique 

### PR DESCRIPTION
**Pull request by DreamWuGit**
when develop unit tests , it is often to manually add RW entries with wrong `rw_counter` specified,  leading to lookup failures eventually, This PR is aim to check `rw_counter ` is unique and increasing at early stage.  it will help trouble shooting this kind of issue for developers.

### CODE REVIEW

**Comment by DreamWuGit** [comment]
@ChihChengLiang @CPerezz @icemelon  Could you please have a look  ?

**Comment by lispc** [comment]
I am a bit confused that "check rw_counter is increasing". You check rw_counter is increasing in **sorted vec**, they always increase ( or, not decease).

If the only purpose is check uniq, then a function like <https://stackoverflow.com/questions/46766560/how-to-check-if-there-are-duplicates-in-a-slice> is more readable than sort and compare

**Comment by DreamWuGit** [comment]
updated to check rw_counter  for now , which is main purpose .

**Comment by ed255** [comment]
What about checking that all the `rw_counter`, once sorted, start at `1` and they increase by `1` for each row?

**Comment by CPerezz** [comment]
Hi @DreamWuGit !

Will review once this is updated to latest `main`

> What about checking that all the `rw_counter`, once sorted, start at `1` and they increase by `1` for each row?

I like that one!

**Comment by DreamWuGit** [comment]
I found some tests'  rw counter not start with 1  and increase by 1 :) will try to modify !

**Comment by DreamWuGit** [comment]
failed tests of slead&sstore refactored in #436 , CallDataSize test refactored in #434 , so all can pass checking now !

**Comment by DreamWuGit** [comment]
> What about checking that all the `rw_counter`, once sorted, start at `1` and they increase by `1` for each row?

Done

File: zkevm-circuits/src/evm_circuit.rs [hunk 1]
```diff
@@ -224,7 +224,22 @@ pub mod test {
...
+                    rows.sort_by_key(|a| a.get_rw_counter());
+                    let mut pre_rw_counter = 0;
+                    for rw in rows {
+                        if pre_rw_counter == 0 {
+                            assert!(rw.get_rw_counter() == 1);
+                        } else {
+                            assert!(rw.get_rw_counter() == pre_rw_counter + 1);
+                        }
+
+                        pre_rw_counter = rw.get_rw_counter();
```
Comment (ed255) [inline zkevm-circuits/src/evm_circuit.rs:242]:
```suggestion
                    let mut expected_rw_counter = 1;
                    for rw in rows {
                        assert!(rw.get_rw_counter() == expected_rw_counter);
                        expected_rw_counter += 1;
```
I believe your implementation is correct, I'm just suggesting a simplification. Feel free to use it or ignore it :)

**Review by ed255 — APPROVED** [review]
I've left a suggestion to simplify the code, but otherwise looks good to me!  These kind of "sanity" checks are always welcome!

File: zkevm-circuits/src/evm_circuit/witness.rs [hunk 2]
```diff
@@ -528,6 +528,20 @@ impl<F: FieldExt> From<[F; 11]> for RwRow<F> {
 }
 
 impl Rw {
+    pub fn get_rw_counter(&self) -> usize {
```
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/witness.rs:531]: it seems changing method name to `rw_counter` will make it more consistent with other methods?

File: zkevm-circuits/src/evm_circuit.rs (hunk 1 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit.rs]: nice to me,  apply it in [2d483e7](https://github.com/appliedzkp/zkevm-circuits/pull/415/commits/2d483e7dc2f5c95d7c911b2243afc4fb6fe8a0d2) thanks

File: zkevm-circuits/src/evm_circuit/witness.rs (hunk 2 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/witness.rs]: nice to me,  apply it in [2d483e7](https://github.com/appliedzkp/zkevm-circuits/pull/415/commits/2d483e7dc2f5c95d7c911b2243afc4fb6fe8a0d2) thanks

**Review by lispc — APPROVED** [review]

**Pull request merged**